pkg/kubelet/pluginmanager/pluginwatcher/monitor.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"context"
+	"os"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"k8s.io/klog/v2"
+	registerapi "k8s.io/kubelet/pkg/apis/pluginregistration/v1"
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/cache"
+	"k8s.io/kubernetes/pkg/kubelet/util"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+const (
+	// dialTimeout is the timeout duration for dialing a plugin socket.
+	dialTimeout time.Duration = time.Second
+
+	// getInfoTimeout is the timeout duration for GetInfo call.
+	getInfoTimeout time.Duration = time.Second
+
+	// pluginListInterval is the duration between getting list of registered plugins from ASW.
+	pluginListInterval time.Duration = time.Second
+
+	// retryInterval is the interval duration between connection checks.
+	connectionCheckInterval time.Duration = time.Second
+
+	// maxFailures is the maximum number of allowed connection monitor failures.
+	// If the number of consecutive failures exceeds this value, the connection is considered dead.
+	maxFailures uint = 2
+)
+
+// pluginConnectionMonitor is a struct to monitor GRPC connections between Kubelet and plugins
+type pluginConnectionMonitor struct {
+	// Directory to watch for plugin sockets
+	sockDir string
+
+	// A sync.Map to store monitored connections
+	connMap sync.Map
+
+	// A channel to stop the monitor
+	stopCh chan struct{}
+	once   sync.Once
+
+	dsw cache.DesiredStateOfWorld
+	asw cache.ActualStateOfWorld
+}
+
+// newPluginConnectionMonitor creates a new plugin connection monitor.
+func newPluginConnectionMonitor(sockDir string, dsw cache.DesiredStateOfWorld, asw cache.ActualStateOfWorld) *pluginConnectionMonitor {
+	return &pluginConnectionMonitor{
+		sockDir: sockDir,
+		connMap: sync.Map{},
+		stopCh:  make(chan struct{}),
+		dsw:     dsw,
+		asw:     asw,
+	}
+}
+
+// start starts the plugin connection monitor.
+func (m *pluginConnectionMonitor) start() {
+	go func() {
+		for {
+			select {
+			case <-m.stopCh:
+				klog.InfoS("Stopping plugin connection monitor")
+				return
+			case <-time.After(pluginListInterval):
+				for _, plugin := range m.asw.GetRegisteredPlugins() {
+					if m.dsw.PluginExists(plugin.SocketPath) {
+						if _, exists := m.connMap.Load(plugin.SocketPath); !exists {
+							m.connMap.Store(plugin.SocketPath, nil)
+							go m.monitorPluginConnection(plugin)
+						}
```

Comment:
what would it mean if it is no in the `connMap`? Just trying to understand what are the else conditions there and whether we need to report them at least in log as an inconsistency?

Reply:
It would mean that this socket is not monitored yet. We don't need to report them as this code checks plugins every  `pluginListInterval` seconds and starts monitoring go-routine only once, when socket is not monitored (yet).

Reply:
I am a bit worried in this and other `else` case that we will have some plugins "stuck" in registration somehow and we will not get any visibility into it. Logging from here will at least tell us that something is wrong

Reply:
Thanks for pointing out! Added more logging.